### ISSUE TITLE
internal namespace is polluted 

### ISSUE THREAD

**Issue by bo0ts** [CGAL/cgal#257]
The internal namespace is polluted which has lead to SFINAE issues (see #129). We need to improve encapsulation between packages by increasing the amount of namespaces.

The guidelines currently advocate adding another namespace `Package` inside internal:

    namespace CGAL { namespace internal { namespace Package {

but I think reversing the order is more natural since we also want to place classes of some packages in a `CGAL::Package` namespace:

   namespace CGAL { namespace Package { namespace internal 

A good starting point to implement this policy would be `Kernel_23` internal global functions.

**Comment by efifogel** [comment]
I believe that the use of the namespace 'internal' was suppose to replace
the use of specific package names as namespaces. So, not only the order
should be replaced, but I find the use of 'internal' redundant in many
cases once the package name is used as a namespace.

**Comment by afabri** [comment]
At one point we discussed to have a namespace per package, so internal should be nested.  We currently only have few of such namespaces, for example for [Surface_mesh_simplification](http://doc.cgal.org/latest/Surface_mesh_simplification/index.html#Chapter_Triangulated_Surface_Mesh_Simplification).

**Comment by sgiraudot** [comment]
Can we carry on this discussion? (Or close the issue if there is nothing to change.)

It seems to me that `internal` and the package namespaces do not have the same aims:
- `internal` should be reserved for non-documented function (this is what internal means, it is used internally by CGAL and is not part of the public API)
- The package namespaces should be used to avoid polluting the CGAL namespaces.

The main issue that I see is: do we keep the current guideline (CGAL/internal/package) or do we switch to what @bo0ts suggested (CGAL/package/internal). This is mainly a developer's issue as the user shouldn't be concerned by the `internal` namespace.

For backwards compatibility, I would choose to keep the current behavior (it seems unrealistic to change this everywhere), but the discussion is still opened.

**Comment by afabri** [comment]
Backward compatibility is no issue as in both cases it is internal code,
I prefer package/internal

**Comment by janetournois** [comment]
I also prefer package/internal. I add the tag dev-meeting so that this issue can be discussed and closed at the next developer meeting.

**Comment by sgiraudot** [comment]
Was this issue discusses at the last developer meeting? If so, what was the decision?

**Comment by lrineau** [comment]
No, it has not been discussed during last developers meeting. Let's hope it will be during the next one.

**Comment by lrineau** [comment]
As we saw recently with a lookup bug in intersection functions, that topic is still an important one. We should do it.

**Comment by MaelRL** [comment]
Was it agreed that `CGAL > Pkg > internal` is to be used everywhere? And are there entities that should be left in `CGAL::internal`?

**Comment by lrineau** [comment]
I agree with the proposal
```C++
CGAL::Pkg::internal
```
instead of the current guidelines
```C++
CGAL::internal::Pkg
```

But I wonder what was the rational of the current guidelines. Does anybody remember?

**Comment by efifogel** [comment]
First, I agree with CGAL::Pkg::internal.
Secondly, I can only guess the origin of CGAL::internal::Pkg.
I think we first had only CGAL::internal, for types and functions that we
didn't want to expose.
Then, instead of reverting everything (as we propose to do now), people
started to add Pkg.

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by afabri** [comment]
My motivation for this PR is Issue #4009 
And I ask package owners as they might be in the middle of a big change, which would mean that I better wait.
